src/plugins/WebOfTrust/util/SynchronousDelayedBackgroundJob.java

```diff
@@ -0,0 +1,231 @@
+     */
+    private void startNewJobThread() {
+        Thread jobThread = new Thread(new Runnable() {
+            @Override
+            public void run() {
+                synchronized(SynchronousDelayedBackgroundJob.this) {
+                    long delay;
+                    // 1. Wait for the deadline to expire
+                    while ((delay = nextRunDeadline - System.currentTimeMillis()) > 0) {
+                        try {
+                            SynchronousDelayedBackgroundJob.this.wait(delay);
+                        } catch (InterruptedException e) {
+                            // Ignore. We can't really return here, since there is at
+                            // least one thread waiting for this thread to complete
+                            // (possibly including side effects caused by its run).
+                        }
+                    }
+                    // 2. Wait until the previous job is done, if any
+                    while (runningJobThread != null) {
+                        try {
+                            SynchronousDelayedBackgroundJob.this.wait();
+                        } catch (InterruptedException e) {
+                            // Ignore. This thread is not even visible from the
+                            // outside at this point, and we will be notified once the
+                            // currently running job is finished. We can't really return
+                            // here, since there is at least one thread waiting for this
+                            // thread to complete (possibly including side effects
+                            // caused by its run).
+                        }
+                    }
+                    // 3. Reset the nextRun variables
+                    nextRunDeadline = NO_DEADLINE;
+                    nextRunScheduled = false;
+                    // From here on, this thread is visible from the outside world,
+                    // and interruption should be handled correctly.
+                    runningJobThread = Thread.currentThread();
+                    // We cannot return here if we are terminating, since that would violate the
+                    // contract specified by {@link #triggerExecution()}. Instead, interrupt the
+                    // thread so that the job may terminate early.
+                    if (isTerminating) {
+                        runningJobThread.interrupt();
+                    }
+                }
+                try {
+                    // 4. Run the background job
+                    job.run();
+                } finally {
+                    synchronized(SynchronousDelayedBackgroundJob.this) {
+                        // 5. Increment runCount
+                        runCount++;
+                        runningJobThread = null;
+                        // 6. Notify all threads waiting upon completion
+                        SynchronousDelayedBackgroundJob.this.notifyAll();
```

Comment:
There is an additional, worse issue to shutdown - mentioning it here instead of at the relevant line 155 so you got it all in one place:
- waitForTermination() will only enter its wait()-loop if isTerminated() detects "!isTerminating || runningJobThread != null" at the time it is called. In the usual case of isTerminating==true, this effectively means that waitForTermination() will NOT enter the wait()-loop if runningJobThread == null.
- runningJobThread will be set in the job thread (created by startNewJobThread()) AFTER the thread has finished waiting for its delay to expire.
- Thus, waitForTermination() will return immediately while a job executor thread is still waiting to execute, as the job executor thread has not set runningJobThread != null yet.
- Also, the caller thread of triggerExecution() will stay in the state of wait()ing for the job thread to execute even after waitForTermination() has finished.

IMHO, it should be implemented to ensure that waitForTermination() returning means:
- no job thread is running OR waiting anymore
- no caller of triggerExecution() is blocked inside that function.

Of course this means that the desire to execute the job once more before shutdown can be ignored.
But lets not forget what our goal was:
- Jobs are for maintenance background jobs. Those shouldn't block shutdown, it should be OK for them to be postponed to the next uptime
- This class was meant for unit tests. They won't even experience the user terminating WOT while they're doing stuff, so terminate() / waitForTermination() won't happen unplanned. If they do happen during unit tests, that will be because the tests specifically try to test stuff about them. In that case, they can be crafted to be OK with the scheduled execution not happening before termination.